---
title: "rna_gfp"
output: html_notebook
date: 2024.12.12
---
##load libraries, read in the sample sheet, assign quant files to obj, construct metadata  

```{r load lib}
library(tidyverse)
library(ggplot2)
library(readr)
library(dplyr)
library(DESeq2)
library(tximeta)
library(vsn)
library(dplyr)
```
#need to point to the dir then read in csv with sample metadata from alignment run. then i need to constuct coldata from metadata in sample info csv by assigning to new obj, then adding cols to coldata by converting df cols to character vectors and concatenating to new df/coldata (adding col data)
```{r load data, construct coldata}
quants_dir <- "/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/rna_egfp/quants"
samp_info <- read.csv("/Users/jbrink/Desktop/20241215_RNAsamplesheetegfp.csv", header = TRUE)

coldata <- samp_info

coldata$names <- paste("1",
                      coldata$SampleID,
                      coldata$Cell.Line,
                      coldata$Treatment,
                      coldata$Replicate,
                      #coldata$Batch,
                      sep = "_"
)

coldata$files <- file.path(quants_dir, coldata$names, "quant.sf")

coldata <- coldata[-c(11, 12, 13), ] #c function returns a vector 

coldata$files
colnames(coldata)[colnames(coldata) == "Cell.Line"] <- "MouseID"
coldata_clean <- coldata[-c(1,2), ]
# Change MouseID for specific rows
#coldata$MouseID[c(1,2)] <- "1723" dropped

coldata_clean$MouseID[c(1,2)] <- "1724"
coldata_clean$MouseID[c(3,4)] <- "1881"
coldata_clean$MouseID[c(5,6)] <- "2139"
coldata_clean$MouseID[c(7,8)]<- "2114"

print(coldata_clean)
```
#next i need to import salmon-generated transcript counts (TPM) and generate a summarized experiment of the counts summarized to gene-level. will also add geneIds
```{r generate summarized experiment and perform some EDA}
sum_exp <- tximeta(coldata_clean)
sum_gene <- summarizeToGene(sum_exp, assignRanges = "abundant")
dds <- DESeqDataSet(sum_gene, design = ~Batch + Treatment)

vsd <- vst(dds)
rownames(vsd)
vsd_mat <- assay(vsd)
vsd_mat
plotPCA(vsd, intgroup = c("Treatment", "Batch"))#PC1 driven by batch. not used in manuscript. pc with batch corrected counts is
```

```{r}
library(pheatmap)
library(RColorBrewer)
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste(vsd$SampleID, vsd$Treatment, sep = " - ")
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)
dds$SampleID
```

```{r tried to sub pc1 for batch effect when dropping sample but this didn't matter}
###########disregard was not used in final model. keeping for notes.
vsd <- vst(dds, blind = TRUE)
pca_data <- prcomp(t(assay(vsd)))
pc1 <- pca_data$x[, 1]


colData(dds)$PC1 <- pc1


design(dds) <- ~ PC1 + Treatment
```

```{r tried to sub pc1 for batch effect when dropping sample but this didn't matter}
#design(dds) = ~Batch + Treatment
des <- DESeq(dds)
vsd <- vst(dds)


results <- results(des)
summary(results)
```

```{r}
vsd_mat <- assay(vsd)
mat_batch <- limma::removeBatchEffect(vsd_mat, batch = vsd$Batch, group = vsd$Treatment)

pca <- prcomp(t(mat_batch))
percent_var <- round(100 * (pca$sdev^2) / sum(pca$sdev^2), 1)

pc_df <- data.frame(
  PC1 = pca$x[, 1],
  PC2 = pca$x[, 2],
  condition = colData(vsd)$Treatment,
  sample = colnames(vsd)
)

ggplot(pc_df, aes(x = PC1, y = PC2, color = condition)) + #pca for paper
  geom_point(size = 3) +  # Scatter plot
  theme_minimal() +
  labs(
    title = "PCA of RNAseq: EGFP+ v EGFP-",
    x = paste0("PC1 (", round(100 * (pca$sdev[1]^2) / sum(pca$sdev^2), 1), "%)"),
    y = paste0("PC2 (", round(100 * (pca$sdev[2]^2) / sum(pca$sdev^2), 1), "%)")
  )
```

```{r more EDA}
#need to do further EDA to see what is driving the effect 
aox4 <- plotCounts(dds, gene = "ENSMUSG00000000567", intgroup = "Treatment") #sox4 is barely expressed across treatment, sox9 expression majorly changes which makes sense
dds

vs_counts <- assay(vsd)
rownames(mat_batch) <- vsd@rowRanges$gene_name
mat_batch
ntd <- normTransform(dds)
meanSdPlot(assay(vsd)) #many lowly expressed genes with no variability filt by mean rank< 10,000 
```

```{r EDA}
gene_var <-rowVars(vs_counts)
log_gene_var <- log1p(gene_var)
hist(log_gene_var)
quantile(log1p(gene_var), probs = c(0.25, 0.5, 0.75, 0.9, 0.95, 0.99, 0.995))
```

```{r}

##didnt use much
threshold <- quantile(log_gene_var, 0.95)
filtered_vs_counts <- vs_counts[log_gene_var > threshold, ]

dim(filtered_vs_counts)
pca <- prcomp(filtered_vs_counts, center = TRUE, scale = TRUE) #pc2 is a batch effect, pc4 is neg v positive effect
pca_df <- pca$rotation

colors <- ifelse(
  coldata$Treatment == "pos",
  "red",
  "black"
)

plot(pca_df [,"PC1"], 
     pca_df[, "PC4"], 
     col = colors, 
     pch = 16
)
```

```{r}
# Extract PCA results for plotting
pca_df <- as.data.frame(pca$rotation)# Scores of the samples
pca_df$Samples <- rownames(pca_df)  # Add sample names

# Plot PC2 vs. PC4 (Batch effect vs. negative/positive effect)
ggplot(pca_df, aes(x = PC1, y = PC2, label = Samples)) +
  geom_point(size = 3, alpha = 0.8) +
  geom_text(vjust = 1.0, hjust = 0.5) +  # Add labels
  theme_minimal() +
  labs(
    title = "PCA Plot",
    x = "PC1",
    y = "PC2"
  )
pca_df$Samples
```

```{r clean up res}
summary(results) #batch + treatment dropped 1723
res_df <- as.data.frame(results)

gene_names <- make.unique(as.character(dds@rowRanges$gene_name))
rownames(res_df) <- gene_names

sorted_by_padj <- res_df[order(res_df$padj, na.last = NA), ]
sorted_by_padj

res_df
```

```{r}
resultsNames(dds)
contrast_res <- results(dds, name = 'PC1')
treat_res <- results(dds, name = "Treatment_pos_vs_neg")
summary(contrast_res)
summary(treat_res)
```

```{r}
library(EnhancedVolcano)

EnhancedVolcano(res_df,
                lab = rownames(res_df),
                x= "log2FoldChange",
                y = "pvalue",
                title = 'Differential Gene Expresion: EGFP+ v EGFP-',
                xlim = c(-5, 5)
)

ggsave(file = "/Users/jbrink/Gracz.sox9/20250303_egfp_deg_volcano_LFC.svg",
        width= 10,
        height= 8
)
```

```{r}
library(msigdbr)
#get hallmark nfkb genes
hallmark_genes_df <- msigdbr(species = "Mus musculus", category = "H")
tnfa_genes <- hallmark_genes_df %>%
  dplyr::filter(gs_name == "HALLMARK_TNFA_SIGNALING_VIA_NFKB") %>%
  pull(gene_symbol)
```

```{r}
gene_names <- make.unique(as.character(dds@rowRanges$gene_name))
rownames(mat_batch) <- gene_names
cts_df <- as.data.frame(mat_batch)
cts_nfkb <- cts_df[tnfa_genes, ]
cts_nfkb <- na.omit(cts_nfkb)
sig_nfkb <- res_df[tnfa_genes,]
sig_nfkb <- na.omit(sig_nfkb)
sig_nfkb


ggplot(sig_nfkb, aes(x = log2FoldChange, y = -log10(padj))) +
  geom_point(aes(color = padj < 0.1), size = 3, alpha = 0.7) +
  geom_text(
    aes(label = ifelse(padj < 0.1, rownames(sig_nfkb), "")),  # Label only significant points
    vjust = 1.5, hjust = 1.5, size = 3, check_overlap = TRUE
  ) +
  geom_hline(yintercept = -log10(0.05), linetype = "dashed", color = "grey") +  # Significance threshold
  geom_vline(xintercept = c(-1, 1), linetype = "dashed", color = "grey") +  # Fold change thresholds
  scale_color_manual(values = c("grey", "red")) +
  labs(
    x = "log2 Fold Change",
    y = "-log10(Adjusted p-value)",
    title = "Volcano Plot of Differentially Expressed NFkB genes",
    color = "Significant (padj < 0.1)"
  ) +
  theme_minimal() +
  theme(
    plot.title = element_text(hjust = 0.5, face = "bold"),
    legend.position = "bottom"
  )
```

```{r save all of these gene lists for plotting with atac}
save_as_csv <- function(dataframe, filename, row.names = TRUE, ...) {
  # Check if the input is a dataframe
  if (!is.data.frame(dataframe)) {
    stop("Input must be a dataframe.")
  }
  
  # Check if the filename is provided and is a character string
  if (!is.character(filename) || length(filename) != 1) {
    stop("Filename must be a single character string.")
  }
  
  # Save the dataframe as a CSV file
  write.csv(dataframe, file = filename, row.names = row.names, ...)
  
  # Print a confirmation message
  message("Dataframe saved as '", filename, "'")
}

save_as_csv(sig_nfkb, "~/Gracz.sox9/src/rna/rnaseq_res_nfkb_gfp.csv")
```

