---
title: "DRAFT_atac_dar_gfp"
output: html_notebook
---
```{r load libraries}
library(DESeq2)
library(plyranges)
library(tidyverse)
library(ComplexHeatmap)
library(msigdbr)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(clusterProfiler)
```

```{r load data and format mat}
load("~/Downloads/20250115ocr_counts.Rdata") 
ocr_bed <- read_bed("/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/atac/union_neg_pos_atac.bed")

ocr_counts$bam.files #need to peform substitution of a matched pattern in a string i.e. bam file names
ms_id <- gsub(".*_(\\d{4}[NP]).*", "\\1", ocr_counts$bam.files) #match characters before target, capture underscore, exactly 4 digits
ms_id

assayNames(ocr_counts)
count_mat <- assay(ocr_counts)
colnames(ocr_counts) <- ms_id

ocr_counts@colData
colnames(ocr_counts)
```

```{r}
ocr_counts@rowRanges
# Exclude rows 48064 and 48065 from rowRanges
filtered_rowRanges <- ocr_counts@rowRanges[-c(48064, 48065)]
granges_df <- as.data.frame(filtered_rowRanges)

# Update the rowRanges slot in ocr_counts
ocr_counts@rowRanges <- filtered_rowRanges
ocr_counts@assays <- ocr_counts@assays[-c(48064, 48065)] #dropped chM rows from counts table

assayNames(ocr_counts)
count_mat <- assay(ocr_counts) #add count data to matrix
colnames(count_mat) <- colnames(ocr_counts) #add mouse id to col names of counts
count_mat <- cbind(count_mat, granges_df)#this count matrix is rows are grange, col are sample and excludes chM
write.csv(count_mat, "/Users/jbrink/Gracz.sox9/src/atac/20250225_counts_matrix_pos_neg_atac.csv") #saving

count_mat
```

######space for me to do some EDA because i found that one of my samples has really low seq depth#####
```{r}
#so i can plot a histogram of counts per peak
dim(count_mat)
total_counts <- rowSums(count_mat) #sum counts across samples per peak
ggplot(data.frame(TotalCounts = total_counts), aes(x = TotalCounts)) +
  geom_histogram(bins = 50, fill = "blue", color = "black") +
  scale_x_log10() + #log trans axis
  labs(title = "distro of total counts per peak",
       x = "total counts (log10)",
       y = "frequency") 
#histogram shows most peaks have low to moderate counts with a few highly accessible regions
```

```{r}
library(reshape)
#box plot 
# Reshape count matrix for ggplot
count_matrix_long  <- melt(count_mat)

# Plot boxplot of counts per sample
ggplot(count_matrix_long, aes(x = variable, y = value)) +
  geom_boxplot(fill = "orange") +
  scale_y_log10() +  # Log-transform y-axis for better visualization
  labs(title = "Distribution of Counts per Sample",
       x = "Samples",
       y = "Counts (log10)") +
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
#box plot shows that sample median count is ~100 per peak
```

```{r}
#need to remove all rows in count mat where peaks are not found in 2/3 samples per condition
# Define replicates for each condition
condition_N <- c("2214N", "2319N", "2341N") 
condition_P <- c("2214P", "2319P", "2341P")

min_count <- 10

# Filter peaks present in at least 2 replicates in either condition
keep <- (rowSums(count_mat[, condition_N] >= min_count) >= 2) |
        (rowSums(count_mat[, condition_P] >= min_count) >= 2)

sum(keep)
filt_peaks_counts <- count_mat[keep, ] #filtered out ~1,000 peaks
```
##end eda
```{r}
dim(filt_peaks_counts)
peak_mat <- filt_peaks_counts[, c("2214N","2214P","2319N", "2319P", "2341N", "2341P")]
samp_info <- data.frame(row.names = colnames(peak_mat),
                        condition = c('neg', 'pos', 'neg', 'pos', 'neg', 'pos'),
                        mouse = c('2214', '2214','2319', '2319', '2341', '2341'),
                        batch = c('A', 'A', 'B', 'B', 'C', 'C') #there has to be a better way to do this 
)
samp_info

dds <- DESeqDataSetFromMatrix(countData = peak_mat,
                              colData = samp_info,
                              design = ~condition)
colData(dds)
```

```{r}
vsd <- vst(dds, blind = TRUE)
pca_data <- prcomp(t(assay(vsd)))
pc1 <- pca_data$x[, 1]


colData(dds)$PC1 <- pc1


design(dds) <- ~ PC1 + condition


dds <- DESeq(dds)
vsd <- vst(dds)
plotPCA(vsd, intgroup = "mouse")

results <- results(dds)
summary(results)
design(dds)
```

```{r}
dds$condition <- relevel(dds$condition, ref = "neg")
des <- DESeq(dds)
res <- results(des)

summary(res)
cooks <- assays(dds)[["cooks"]]

par(mar=c(8,5,2,2))
boxplot(log10(cooks), range=0, las=2)
```

```{r}
results(des)

res_shk <- lfcShrink(des, 
                      coef = "condition_pos_vs_neg", 
                      type = "apeglm"
)
res_shk
res_shk_cts <- cbind(res_shk, filt_peaks_counts)
write.csv(res_shk_cts,"/Users/jbrink/Gracz.sox9/src/atac/20250225_res_cts_matrix_pos_neg_atac.csv")



vsd <- vst(dds)
plotPCA(vsd, intgroup = "mouse")
mat <- assay(vsd)
mat_batch4 <- limma::removeBatchEffect(mat, covariates = vsd$PC1, group = vsd$condition)
dim(mat_batch4)
head(mat_batch4)
```

```{r}
library(ChIPpeakAnno)
library(ChIPseeker)

res_bed <- filt_peaks_counts[, !colnames(filt_peaks_counts) %in% c("2214N", "2214P", "2319N", "2319P", "2341N", "2341P")]
res_bed <- makeGRangesFromDataFrame(res_bed, keep.extra.columns = TRUE)
peak_anno <- annotatePeak(res_bed, 
                                TxDb = TxDb.Mmusculus.UCSC.mm10.knownGene, 
                                annoDb = "org.Mm.eg.db",
                                tssRegion = c(-3000, 3000)
)
peak_anno
grange_anno <- as.GRanges(peak_annotation)
```

```{r not all peaks were annotated and needed to find them in bed file then remove}
#in new bed where peaks were dropped that were not found in 2/3 bio reps this is no longer an issue
findOverlaps(grange_anno, ocr_bed)

diff_range <- GenomicRanges::setdiff(ocr_bed, grange_anno)

print(diff_range) #these are chrM regions that werent annotated in the peak_anno. so they must go
ocr_bed <- dropSeqlevels(ocr_bed, value = "chrM", pruning.mode = "coarse") #i already dropped these from the counts data during a cleaning step
```

```{r}
#merged_data <- merge(ocr_bed, deseq2_results, by = "peak_id")
```


```{r add anno cols to counts}
peak_meta <- mcols(peak_anno@anno)
peak_meta
vsd
vst_mat <- assay(vsd)
meta_cols <- c("annotation", "SYMBOL", "geneChr", "geneStart", "geneEnd")
vst_meta <- cbind(peak_meta[, meta_cols], mat_batch4)
vst_meta <- vst_meta[!grepl("^Gm|Rik$", vst_meta$SYMBOL), ]
vst_meta
base::saveRDS(vst_meta, "vst_meta.rds")
```

```{r}
# Extract the last 6 columns
counts_matrix <- as.matrix(vst_meta[, 6:11])

# Calculate the standard deviation for each row
row_std_dev <- apply(counts_matrix, 1, sd)

# Add the standard deviation to the DataFrame
vst_meta$std_dev <- row_std_dev

# Sort by standard deviation and select top 1500 regions
top_regions <- vst_meta[order(vst_meta$std_dev, decreasing = TRUE), ]
top_regions <- top_regions[1:4000, ]

# Create a BED file from the top regions
top_regions_bed <- data.frame(
  chr = top_regions$geneChr,
  start = top_regions$geneStart,
  end = top_regions$geneEnd
)

# Save to a BED file
write.table(top_regions_bed, "top_4000_variable_regions.bed", sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

```{r}
dim(mat_batch)
# Calculate the standard deviation for each row
batch_std_dev <- apply(mat_batch, 1, sd)

# Add the standard deviation to the DataFrame
mat_batch$std_dev <- batch_std_dev

head(mat_batch)

# Sort by standard deviation and select top 1500 regions
top_regions <- mat_batch[order(mat_batch$std_dev, decreasing = TRUE), ]
top_regions <- top_regions[1:4000, ]
```


```{r}
pca <- prcomp(t(mat_batch4))
percent_var <- round(100 * (pca$sdev^2) / sum(pca$sdev^2), 1)

pc_df <- data.frame(
  PC1 = pca$x[, 1],
  PC2 = pca$x[, 2],
  condition = colData(vsd)$condition,
  sample = colnames(vsd)
)

pca <- ggplot(pc_df, aes(x = PC1, y = PC2, color = condition, label = sample)) +
  geom_point(size = 3) +  # Scatter plot
  geom_text(vjust = 1.0, hjust = 1.0) +  # Add sample labels
  theme_minimal() +
  labs(
    title = "PCA of ATACseq Samples",
    x = paste0("PC1 (", round(100 * (pca$sdev[1]^2) / sum(pca$sdev^2), 1), "%)"),
    y = paste0("PC2 (", round(100 * (pca$sdev[2]^2) / sum(pca$sdev^2), 1), "%)")
  )
pca
ggsave(pc_df, "/Users/jbrink/Gracz.sox9/pca_atac_neg_pos.svg")
```

```{r}
vst_meta
vst_meta[, "annotation_summary"] <- str_split_i(vst_meta$annotation, " ", 1)

res_df <- as.data.frame(res)
summary(res)

res_meta <- cbind(res_df, peak_meta)
res_meta[, "annotation_summary"] <- str_split_i(res_meta$annotation, " ", 1)
res_meta

res_clean <- res_meta[, -7] # Drop the name col of NA


res_clean <- res_clean %>%
  mutate(X1 = paste0("chr", X1))

res_clean <- res_clean[!grepl("^Gm|Rik$|Mup$", res_clean$SYMBOL), ]
res_clean

# Remove rows with NA values in padj or log2FoldChange
res_clean <- res_clean %>%
  filter(!is.na(padj) & !is.na(log2FoldChange))
```

```{r}
library(EnhancedVolcano)
EnhancedVolcano(res_clean,
                lab = res_clean$SYMBOL,
                x = 'log2FoldChange',
                y = 'pvalue')

```


```{r}
res_clean

padj_threshold <- 0.1
log2FC_threshold <- 0.5

# Now filter for significant regions
up_regulated <- res_clean %>%
  filter(padj < padj_threshold & log2FoldChange > log2FC_threshold)

down_regulated <- res_clean %>%
  filter(padj < padj_threshold & log2FoldChange < -log2FC_threshold)
```

```{r}
# Load dplyr
library(dplyr)

# Rename columns
up_regulated <- up_regulated %>%
  rename(
    chr = geneChr,
    start = geneStart,
    end = geneEnd
  )

down_regulated <- down_regulated %>%
  rename(
    chr = geneChr,
    start = geneStart,
    end = geneEnd
  )
```

```{r}
# Function to save data as BED file
save_as_bed <- function(data, filename) {
  bed_data <- data %>% select(chr, start, end)
  write.table(bed_data, file = filename, sep = "\t", row.names = FALSE, col.names = FALSE, quote = FALSE)
}

# Save up-regulated regions
save_as_bed(up_regulated, "up_regulated.bed")

# Save down-regulated regions
save_as_bed(down_regulated, "down_regulated.bed")
colnames(down_regulated)
```

```{r}
plot_cts <- mat_batch4[order(res_clean$padj)[1:138], ]
plot_cts <- mat_batch4[order(res_clean$log2FoldChange), ]
plot_cts <- as.data.frame(plot_cts)

plot_cts$SYMBOL <- make.unique(plot_cts$SYMBOL)

set.seed(123)
pheatmap(as.matrix(plot_cts[, 4:ncol(plot_cts)-1]),
         cluster_rows = TRUE,       
         cluster_cols = TRUE,           
         scale = "row",
         color = colorRampPalette(c("blue", "white", "red"))(50),
         main = "Heatmap of Significant DARs")
plot_cts
```

```{r}
res_meta <- res_meta[!grepl("^Gm|Rik$|Mup$", res_meta$SYMBOL), ]
res_meta$significant <- with(res_meta, padj < 0.1 & abs(log2FoldChange) > 1)

ggplot(res_meta, aes(x = log2FoldChange, y = -log10(padj), color = significant)) +
  geom_point(alpha = 0.6, size = 1.5) +         # Points with transparency
  scale_color_manual(values = c("grey", "black")) +  # Color significant points
  geom_text(
    data = subset(res_meta, significant),            # Label significant points
    aes(label = SYMBOL),
    size = 3, hjust = 0, vjust = 1, check_overlap = TRUE
  ) +
  theme_minimal() +                              # Clean theme
  labs(
    title = "Volcano Plot of all DAR",
    x = "Log2 Fold Change",
    y = "-log10(Adjusted P-value)",
    color = "Significant"
  ) +
  theme(legend.position = "top")
)
```

```{r}
plotAnnoPie(peak_anno)
```

```{r goterm of intergenic, intronic, promoter elements}
head(res_clean)
res_clean <- res_meta[, -6]
res_clean

res_promoter <- res_clean %>% filter(str_detect(annotation_summary, "Promoter"))
res_intron <- res_clean %>% filter(str_detect(annotation_summary, "Intron"))
res_distal <- res_clean %>% filter(str_detect(annotation_summary, "Distal"))

saveRDS(res_promoter, file = "res_promoter.rds")
saveRDS(res_intron, file = "res_intron.rds")
saveRDS(res_distal, file = "res_distal.rds")
```
