---
title: "DEG_uninj_inj_Sox9_Heps"
author: "Jackie Brinkman"
format: html
---
## Quarto
```{r load libraries}
#| echo: true
#| output: false

library(tidyverse)
library(ggplot2)
library(DESeq2) 
library(tximeta)
library(readr)
library(devtools)
library(BiocManager)
library(EnhancedVolcano)
library(pheatmap)
library(ComplexHeatmap)
library(renv)
```
```{r load data}
#point to quants files 
input_dir <- "/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/rna/quants"
#point to sample metadata 
samples <- read.csv('/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/20240320_RNAsamplesheet.csv', header=TRUE)
```

```{r coldata metadata}
#| output: false
#make coldata by copying sample metadata, adding col of files, names (which is sampleID)
coldata <- samples
coldata$names <- paste('1',
                       coldata$SampleID, 
                       coldata$Cell.Line, 
                       coldata$Treatment, 
                       coldata$Replicate, 
                       #coldata$Batch,
                       sep = '_')
#file.path basically constructs the path & produces a character vector of args concat term by term
coldata$files <- file.path(input_dir, coldata$names, 'quant.sf')
```

```{r don't run. for removing outlier 1638 and testing model. outlier removal is negligable}
#row_to_keep = c(TRUE, FALSE, TRUE, TRUE, TRUE, TRUE, TRUE, TRUE)
#newdata <- coldata[row_to_keep, ]
```


```{r coldata metadata}
#generate RangedSummarizedExperiment
se <- tximeta(coldata)
#include assignRanges controls the way that the rowRanges of the output object are assigned. choose abundant for most abundant isoform (expressed gene)
gse <- summarizeToGene(se, assignRanges = "abundant")
```
```{r Make DESeqDataset}
gse$Batch <- factor(gse$Batch)
#Make DESeqDataset. filter out lowly expressing genes in dds then vst the counts in dds
dds <- DESeqDataSet(gse, design = ~ Batch + Treatment)
dds$Treatment <- relevel(dds$Treatment, ref = "chow")

#generate count matrix
counts <- assay(gse)
keep <- rowSums(counts(dds)) >= 25
dds <- dds[keep,]
vsd <- vst(dds)

# get vs scaled counts and set rownames to gene symbol
vs_counts <- assay(vsd)
rownames(vs_counts) <- rowData(vsd)$gene_name

#write counts x gene to csv
counts_genes_df_csv <- write.csv(vs_counts, 
                              file ='/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/20241010_uninj_inj_counts.csv')
```

```{r make a pca of samples after batch effect removal for paper viz}
mm <- model.matrix(~Treatment, colData(vsd))
mat <- limma::removeBatchEffect(vs_counts, batch=vsd$Batch, design=mm)
assay(vsd) <- mat
plotPCA(vsd, 
        intgroup = c("Treatment", "SampleID")) + 
        ggplot2::theme_classic() + 
        scale_color_brewer(palette = "Paired")
```

```{r}
#commenting out b/c I wanted a plot that removed batch effect var for publication viz 
#plotPCA(vsd, 
        #intgroup = c("Treatment", "SampleID"))+ ggplot2::theme_classic() + scale_color_brewer(palette = "Paired")
```
```{r DESeq2 linear modeling}
dds <- DESeq(dds)
resultsNames(dds)
```
```{r DESeq2 linear modeling results}
res <- results(dds, tidy = TRUE) #setting col and then converting to a tibble
#summary(res) shows there are 12 NA's in padj column and 0 across other stat cols. this is fine.
```

```{r just making sure that there are no overlapping NAs}
na_both_count <- sum(is.na(res_shrink2$padj)) #& is.na(res_shrink$log2FoldChange))

print(paste("Number of rows where both padj and log2FoldChange are NA:", na_both_count))
```
```{r box plot for cooks distance b/c was having trouble with outliers and wanted to check samples}
par(mar=c(8,5,2,2))
boxplot(log10(assays(dds)[["cooks"]]), range=0, las=2)
```
```{r DESeq2 res df add gene symbols}
#build table of additional gene symbols has the first column called "row"
res_gene_names <- data.frame(row = res$row, 
                             symbol = rowData(dds)$gene_name)


res_genes_df <- merge(res, 
                      res_gene_names,
                      all.x = TRUE)
#write results
res_genes_df_csv <- write.csv(res_genes_df, 
                              file ='/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/20241010_uninj_inj_DEGFig3.csv')
```

```{r DESeq2 res shrink LFC}
resShrink <- lfcShrink(dds,  
                       coef = "Treatment_6wk.ddc_vs_chow",
                       type="apeglm"
)

tidy_resShrink <- as.data.frame(resShrink) 
#assign symbol in new.table to new col in tidy_resShrink
tidy_resShrink$symbol <- res_gene_names$symbol

#write results
res_LFCshrink_genes_df_csv <- write.csv(res_genes_df, 
                                      file='/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/20241010_uninj_inj_DEG_LFCshrinkFig3.csv')
```

```{r}
# Identify Mup genes in tidy_resShrink
is_mup <- str_starts(tidy_resShrink$symbol, "Mup")

# Filter out Mup genes from tidy_resShrink
tidy_resShrink_no_mups <- tidy_resShrink[!is_mup, ]
write.csv(tidy_resShrink_no_mups, file='/Users/jbrink/Desktop/Lab2024/gracz.sox9_archive/20241010_DEG_FIG3_noMups_shrunkres.csv')
                                  
```

```{r volcano plot of all genes with LFC shrink}
EnhancedVolcano(tidy_resShrink_no_mups,
                lab = tidy_resShrink_no_mups$symbol,
                x = 'log2FoldChange',
                y = 'pvalue',
                title = 'DEG Sox9+ hepatocytes: injured v. uninjured',
                #selectLab = c('Bmp6','Klf4'),
               # pCutoff = 10e-30,
                #FCcutoff = 0.5,
                pointSize = 3.0,
                labSize = 6.0
)
ggsave(file = "/Users/jbrink/Gracz.sox9/20241010_USETHISFIG3_uninj_inj_volcano_LFCshrinkres.svg",
        width= 10,
        height= 8
)
```
```{r data viz. all genes}
#rownames(vs_counts) <- rowData(vsd)$gene_name as a reminder what vs_counts is
select <- order(rowMeans(vs_counts), decreasing=TRUE)[1:30]
df <- coldata[, c("Treatment", "SampleID")]
rownames(df) <- colnames(vs_counts)

pheatmap(vs_counts[select,], 
         cluster_rows=FALSE, 
         show_rownames=TRUE, 
         show_colnames = FALSE,
         cluster_cols=FALSE,
         annotation_col=df)
```
```{r filtering out non-sig genes from results before plotting}
res_filt <- tidy_resShrink_no_mups[abs(tidy_resShrink_no_mups$log2FoldChange) > 4.0 & tidy_resShrink_no_mups$padj < 0.1, "symbol"]
#res_filt <- tidy_resShrink[abs(tidy_resShrink$log2FoldChange) > 2.0, "symbol"]
```


```{r filtering out non-sig genes from results before plotting}
# get mup genes and remove them 
is_mup <- str_starts(rownames(vs_counts), "Mup") # returns logic vector
mups <- rownames(vs_counts)[is_mup]# contains mup gene symbols
counts_no_mups <- vs_counts[!is_mup,]
#res_no_mups <- res_filt[!res_filt %in% mups]
```

```{r filtering out non-sig genes from results before plotting}
# Check for missing genes
missing_genes <- res_filt[!res_filt %in% rownames(vs_counts)]
if (length(missing_genes) > 0) {
  warning("The following genes are not present in vs_counts: ", paste(missing_genes, collapse = ", "))
}

#this shows that filtering for padj means leaving 11 NAs. I need to filter them out.
# Remove rows with NA values from topDE
res_filt <- na.omit(res_filt)

# counts of topDE genes with mups removed
plot_genes <- vs_counts[res_filt,]
```

```{r plotting specific gene counts}
plotCounts(dds, gene= "ENSMUSG00000037852", intgroup = "Treatment")
```
```{r}
# column annotations for pheatmap
heat_annot <- coldata[, c("Treatment", "SampleID")]
rownames(heat_annot) <- colnames(vs_counts)

pheatmap(plot_genes, 
         cluster_rows=FALSE, 
         show_rownames=FALSE, 
         show_colnames = FALSE,
         cluster_cols=FALSE,
         annotation_col=heat_annot,
         scale = "row")
```
```{r trying other heatmaps for clustering analysis}
mat <- vs_counts[res_filt,]
mat <- mat - rowMeans(mat)

gene_clust <- Heatmap(mat,
                      row_km = 4,
                      show_row_names = FALSE,
                      show_column_names = TRUE,
                      show_column_dend = FALSE,
                      )
clust_list <- row_dend(gene_clust)
```

```{r}
#kmeans clustering of DEG: LFC abs(4) and padj <0.05 =193 genes
set.seed(123)
hm <- Heatmap(mat,
        row_km = 4,
        row_km_repeats = 2,
        show_row_names = FALSE,
        show_column_names = TRUE,
        show_column_dend = FALSE,
        show_row_dend = FALSE,
        )
hm
```

```{r}
# Set the output to an SVG file
svg("20241010_DEGFIG3_kmeans_heatmap_output.svg", width = 10, height = 8) 
set.seed(123)
hm <- draw(hm)


gene_clusts <- row_dend(hm)

clust1_genes <- row_order(hm)[['4']]
```

```{r}
mat[clust1_genes, ]

rownames(mat[clust1_genes, ])
```

```{r}
clusters <- gene_clust$kmeans

mat2 <- as.data.frame(select2)
mat2$clusters <- clusters$cluster
write.csv(mat2, "/Users/jbrink/Desktop/Lab2024/gracz.sox9/rna/RNA_uninjured_injured_DEGkmeans.csv")
```

```{r data viz. GSEA}
#uninjured v DDC hepatocyte DEG enriched for pro-inflammatory gene regulation Geneset
library(enrichplot)
library(ggplot2)
library(clusterProfiler)
library(org.Mm.eg.db)
library(msigdbr)
library(GSEABase)
library(msigdb)

m_df <- getMsigdb(
  org = "mm",
  id = "SYM"
)

h_pathways <- msigdb::subsetCollection(
  m_df,
  collection = "h"
)

#this is a function made by CY#
make_gsea_term_df <- function(path_col) {
  pathways <- Map(str_to_upper, geneIds(path_col))
  df <- stack(pathways)
  df <- df[, c("ind", "values")]
  colnames(df) <- c("pathway", "symbol")
   return(df)
}

h_paths <- make_gsea_term_df(h_pathways)
h_gene_sets <- split(h_paths$symbol, h_paths$pathway)
h_gene_sets_df <- stack(h_gene_sets)
colnames(h_gene_sets_df) <- c("GENE", "TERM")

h_gene_sets_df <- data.frame(
  TERM = as.character(h_gene_sets_df$TERM),
  GENE = as.character(h_gene_sets_df$GENE)
)

# get differential expression results from DESeq2
deseq_res <- results(dds)

# Create a ranked gene list by log2 fold change
ranked_genes <- tidy_resShrink_no_mups$log2FoldChange
names(ranked_genes) <- tidy_resShrink_no_mups$symbol
names(ranked_genes) <- toupper(names(ranked_genes))
ranked_genes <- ranked_genes[!is.na(ranked_genes)]
ranked_genes <- sort(ranked_genes, decreasing = TRUE)

ranked_genes <- ranked_genes[names(ranked_genes) %in% h_gene_sets_df$GENE]


gsea_res <- GSEA(
  ranked_genes,
  TERM2GENE = h_gene_sets_df,
  pvalueCutoff = 0.05,
  verbose = TRUE
)
```
```{r}
head(gsea_res@result)
dotplot(gsea_res, showCategory = 15)
```
```{r}
gseaplot(gsea_res, geneSetID = "HALLMARK_IL6_JAK_STAT3_SIGNALING")
```